docs/async.md

```diff
@@ -0,0 +1,210 @@
+
+### Making requests
+
+The request methods are all async, so you should use `response = await client.get(...)` style for all of the following:
+
+* `AsyncClient.get(url, ...)`
+* `AsyncClient.options(url, ...)`
+* `AsyncClient.head(url, ...)`
+* `AsyncClient.post(url, ...)`
+* `AsyncClient.put(url, ...)`
+* `AsyncClient.patch(url, ...)`
+* `AsyncClient.delete(url, ...)`
+* `AsyncClient.request(method, url, ...)`
+* `AsyncClient.send(request, ...)`
+
+### Opening and closing clients
+
+Use `async with httpx.AsyncClient()` if you want a context-managed client...
+
+```python
+async with httpx.AsyncClient() as client:
+    ...
+```
+
+Alternatively, use `await clien.aclose()` if you want to close a client explicitly.
+
+```python
+client = httpx.AsyncClient()
+...
+await client.aclose()
+```
+
+### Streaming responses
+
+The `AsyncClient.stream(method, url, ...)` method is an async context block.
+
+```python
+>>> client = httpx.AsyncClient()
+>>> async with client.stream('GET', 'https://www.example.com/') as response:
+>>>     async for chunk in r.aiter_content():
+>>>         ...
+```
+
+The async response streaming methods are:
+
+* `Response.aread()` - For conditionally reading a response inside a stream block.
+* `Response.aiter_content()` - For streaming the response content as bytes.
+* `Response.aiter_text()` - For streaming the response content as text.
+* `Response.aiter_lines()` - For streaming the response content as lines of text.
+* `Response.aiter_raw()` - For streaming the raw response bytes, without applying content decoding.
+* `Response.aclose()` - For closing the response. You don't usually need this, since `.stream` block close the response automatically on exit.
+
+### Streaming requests
+
+When sending a request body with an `AsyncClient` instance, it is invalid to use a bytes generator, but it is valid to use an async bytes generator...
+
+```
+async def upload_bytes():
+    ...  # yield byte content
+
+client.post(url, data=upload_bytes())
+```
+
+## Supported async environments
+
+HTTPX supports either `asyncio` or `trio` as an async environment.
+
+By default it will auto-detect which of those two to use as the backend
+for socket operations and concurrency primitives.
+
+You can also explicitly select a backend by instantiating a client with the
+`backend` argument...
+
+```python
+client = httpx.AsyncClient(backend='auto')     # Autodetection. The default case.
+client = httpx.AsyncClient(backend='asyncio')  # Use asyncio as the backend.
+client = httpx.AsyncClient(backend='trio')     # Use trio as the backend.
+```
+
+### [AsyncIO](https://docs.python.org/3/library/asyncio.html)
+
+AsyncIO is Python's [built-in library](https://docs.python.org/3/library/asyncio.html)
+for writing concurrent code with the async/await syntax.
+
+```python
+import asyncio
+import httpx
+
+async def main():
+    client = httpx.AsyncClient()
+    response = await client.get('https://www.example.com/')
+    print(response)
```

Comment:
Shouldn't we be encouraging the `async with` syntax here?

I'm not even sure why this runs fine without a "an open stream is being garbage collected" error, since the keepalive connection should still be open and have to be closed off. It does raise a warning:

```console
$ python -Wd debug/example.py
<Response [200 OK]>
sys:1: ResourceWarning: unclosed <socket.socket fd=9, family=AddressFamily.AF_INET6, type=SocketKind.SOCK_STREAM, proto=6, laddr=('2a01:cb04:49:9800:4128:7fd2:96fd:60d0', 50895, 0, 0), raddr=('2606:2800:220:1:248:1893:25c8:1946', 443, 0, 0)>
ResourceWarning: Enable tracemalloc to get the object allocation traceback
/Users/florimond/.pyenv/versions/3.8.0/lib/python3.8/asyncio/sslproto.py:321: ResourceWarning: unclosed transport <asyncio.sslproto._SSLProtocolTransport object at 0x103656160>
  _warn(f"unclosed transport {self!r}", ResourceWarning, source=self)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
/Users/florimond/.pyenv/versions/3.8.0/lib/python3.8/asyncio/selector_events.py:684: ResourceWarning: unclosed transport <_SelectorSocketTransport fd=9>
  _warn(f"unclosed transport {self!r}", ResourceWarning, source=self)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
```

(*We* may know that there are *some* cases when it just happens to be fine to not close the client — in this case it's not — but I really think we should encourage users to always ensure the client gets closed at some point.)

Reply:
Yeah good call.